lib/scholar/manifold/tsne.ex

```diff
@@ -0,0 +1,297 @@
+      default: :squared_euclidean,
+      doc: ~S"""
+      Metric used to compute the distances.
+      """
+    ],
+    exaggeration: [
+      type: {:or, [:float, :pos_integer]},
+      default: 10.0,
+      doc: ~S"""
+      Controls how tight natural clusters in the original space are in the embedded space and
+      how much space will be between them. For larger values, the space between natural clusters will be larger in the embedded space.
+      """
+    ]
+  ]
+
+  @opts_schema NimbleOptions.new!(opts_schema)
+
+  @doc """
+  Fits a PCA for sample inputs `x`.
+
+  ## Options
+
+  #{NimbleOptions.docs(@opts_schema)}
+
+  ## Return Values
+
+  Returns the embedded data `y`.
+
+  ## Examples
+
+      iex> x = Nx.iota({4,5})
+      iex> seed = 42
+      iex> Scholar.Manifold.TSNE.fit(x, num_components: 2, seed: seed)
+      #Nx.Tensor<
+        f32[4][2]
+        [
+          [1428.49609375, 0.0],
+          [-1499.8349609375, 0.0],
+          [-418.6555480957031, 0.0],
+          [490.010009765625, 0.0]
+        ]
+      >
+  """
+  deftransform fit(x, opts \\ []) do
+    opts = NimbleOptions.validate!(opts, @opts_schema)
+    seed = Keyword.get_lazy(opts, :seed, &System.system_time/0)
+    fit_n(x, seed, opts)
+  end
+
+  defnp fit_n(x, seed, opts \\ []) do
+    {perplexity, learning_rate, num_iters, num_components, exaggeration, init, metric} =
+      {opts[:perplexity], opts[:learning_rate], opts[:num_iters], opts[:num_components],
+       opts[:exaggeration], opts[:init], opts[:metric]}
+
+    x = to_float(x)
+    {n, _dims} = Nx.shape(x)
+
+    y =
+      case init do
+        :random ->
+          key = Nx.Random.key(seed)
+
+          {y, _new_key} =
+            Nx.Random.normal(key, 0.0, 1.0e-4, shape: {n, num_components}, type: Nx.type(x))
+
+          y
+
+        :pca ->
+          Scholar.Decomposition.PCA.fit_transform(x, num_components: num_components)
+      end
+
+    ys = Nx.broadcast(0.0, {num_iters, n, num_components})
+
+    ys = Nx.put_slice(ys, [0, 0, 0], Nx.new_axis(y, 0))
+    ys = Nx.put_slice(ys, [1, 0, 0], Nx.new_axis(y, 0))
+
+    p = p_joint(x, perplexity, metric)
+
+    {y, _, _, _, _} =
+      while {_y = y, ys, learning_rate, p, i = 2}, i < num_iters do
+        q = q_joint(Nx.take(ys, i - 1), metric)
+        grad = gradient(p * exaggeration(i, exaggeration), q, Nx.take(ys, i - 1), metric)
+
+        temp =
+          Nx.take(ys, i - 1) - learning_rate * grad +
+            +momentum(i) * (Nx.take(ys, i - 1) - Nx.take(ys, i - 2))
+
+        ys = Nx.put_slice(ys, [i, 0, 0], Nx.new_axis(temp, 0))
+        {temp, ys, learning_rate, p, i + 1}
+      end
+
+    y
+  end
+
+  defnp pairwise_dist(x, metric) do
+    {num_samples, num_features} = Nx.shape(x)
+
+    t1 = Nx.new_axis(x, 0) |> Nx.broadcast({num_samples, num_samples, num_features})
+    t2 = Nx.new_axis(x, 1) |> Nx.broadcast({num_samples, num_samples, num_features})
```

Comment:
```suggestion
    t1 = x |> Nx.reshape({1, num_samples, num_features) |> Nx.broadcast({num_samples, num_samples, num_features})
    t2 = x |> Nx.reshape({num_samples, 1, num_features) |> Nx.broadcast({num_samples, num_samples, num_features})
```
I'd rather use reshape here because it makes it easier to visualize what the broadcast is doing.

Do you need the explicit broadcasting here, btw?

Reply:
Yeah, I need explicit broadcasting since Scholar.Metric.Distance expects t1 and t2 to have the same shapes.